### PR TITLE
Unit test controller and service of each core Module

### PR DESCRIPTION
**Pull request by LeonFLK**
## fixes KILTProtocol/ticket#616
##fixes KILTProtocol/ticket#721
##fixes KILTProtocol/ticket#720
##fixes KILTProtocol/ticket#718
##fixes KILTProtocol/ticket#686
##fixes KILTProtocol/ticket#724
##fixes KILTProtocol/ticket#730

- [x] CTypes module
- [x] Messaging Module
- [x] Contacts Module
- [x] Faucet Module

This introduces unit testing of controller and service for each core module, via faking the modules service and mocking of the mongoose Models.
Removes naive approach of in-memory unit testing.
Also aims to refactor some minor issues within our classes.

Contact Module changes:
Removed _findByAddress, and replaced calling it with directly querying the DB.
converToContact is now a private function inside MongoDbMContactsService.

CType Module changes:
Renamed verifyCType to verifyCTypeAndReturnChainOwner, to more accurately represent the updated behaviour. Please suggest a better name!
Removed _findByHash, replaced calls with directly querying DB.
Changed convertToCType from exported to private.

Faucet Module Changes:
Removed email from POST payload parameter.
Removed try catch to check address in favor of checkAddress call.
Now only accepts actual addresses and NOT public keys, or anthing else, prefix of address has to match KILT prefix of 42.

Messaging Module Changes:
Changed uuidv4() from unexported function to exported constant.

## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

**Review by rflechtner — CHANGES_REQUESTED** [review]
Be sure to `await` any `expect(async() => {}).resolves` and `expect(async() => {}).rejects` like this: 
`await expect(async() => true).resolves.toBeTruthy()`
Other than that I have no functional complaints, the remaining comments are all style-related. To summarise, there's two things that could help readability of the tests:
1. Choose one way to mock async functions and to return promises. My personal favourite is `async(): Promise<any> => value`, i.e. `async(): Promise<void> => {}`. `(): Promise<any> => Promise.resolve(value)` also works, but `async (): Promise<any> => Promise.resolve(value)` seems a bit redundant
2. Separate negative and positive tests. In general its nicer not to lump too many different cases in one `it()`, but checking happy path & correct error handling should definitely be two tests imo

File: src/ctypes/ctypes.module.spec.ts [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+      })
+    })
+  })
+  class CTypeModel {
+    public static find = jest
+      .fn()
+      .mockReturnValue({ exec: () => [] as CTypeDB[] })
+    public static findOne = jest
+      .fn()
+      .mockReturnValue({ exec: () => Optional.ofNullable(null) })
```
[the remark this answers is not in the document]
Comment (LeonFLK) [reply src/ctypes/ctypes.module.spec.ts:232]: yes

File: src/contacts/contacts.module.spec.ts [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        addSpy.mockClear()
+        await expect(contactsController.add(noAddressContact)).rejects.toThrow(
+          BadRequestException
+        )
+        await expect(contactsController.add(noPubKeyContact)).rejects.toThrow(
+          BadRequestException
+        )
+        await expect(contactsController.add(noNameContact)).rejects.toThrow(
+          BadRequestException
+        )
```
[the remark this answers is not in the document]
Comment (LeonFLK) [reply src/contacts/contacts.module.spec.ts:159]: Yeah for mopst tests I like to have them purely separated, but for this one it jest depends on one missing property, so I think these negative tests are all somehow related, as they trigger the same response from the controller.

File: src/contacts/contacts.module.spec.ts [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+        expect(mockedHashStr).toHaveBeenCalledTimes(0)
+        expect(mockedVerify).toHaveBeenCalledTimes(0)
+        expect(addSpy).toHaveBeenCalledTimes(0)
```
[the remark this answers is not in the document]
Comment (LeonFLK) [reply src/contacts/contacts.module.spec.ts:162]: Yes, explicitly stating that the function should not have been called is great!

File: src/contacts/contacts.module.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rflechtner) [reply src/contacts/contacts.module.spec.ts:166]: true, but that's a reason to group the negative tests, not one to lump them together with the positive test. What I meant was move them out of this test, like so: 
```typescript
      it('calls contactService.add on valid Contact without Did', async () => {
        const addSpy = jest.spyOn(contactsService, 'add')
        expect(await contactsController.add(testContact)).toEqual(undefined)
        expect(mockedHashStr).not.toHaveBeenCalled()
        expect(mockedVerify).not.toHaveBeenCalled()
        expect(addSpy).toHaveBeenCalledTimes(1)
        expect(addSpy).toHaveBeenCalledWith(testContact)
        addSpy.mockClear()
      })
      it('rejects contact if property missing', async () => {
        const addSpy = jest.spyOn(contactsService, 'add')
        const noAddressContact: Contact = {
          ...testContact,
          publicIdentity: { ...testContact.publicIdentity, address: '' },
        }
        const noPubKeyContact: Contact = {
          ...testContact,
          publicIdentity: {
            ...testContact.publicIdentity,
            boxPublicKeyAsHex: '',
          },
        }
        const noNameContact: Contact = {
          ...testContact,
          metaData: {
            ...testContact.metaData,
            name: '',
          },
        }
        await expect(contactsController.add(noAddressContact)).rejects.toThrow(
          BadRequestException
        )
        await expect(contactsController.add(noPubKeyContact)).rejects.toThrow(
          BadRequestException
        )
        await expect(contactsController.add(noNameContact)).rejects.toThrow(
          BadRequestException
        )
        expect(mockedHashStr).not.toHaveBeenCalled()
        expect(mockedVerify).not.toHaveBeenCalled()
        expect(addSpy).not.toHaveBeenCalled()
      })
```

[the remark this answers is not in the document]
Comment (rflechtner) [reply src/contacts/contacts.module.spec.ts:166]: btw you may be able to replace all those `spy.mockClear()` with a call in `beforeAll` to [clear/reset all mocks](https://jestjs.io/docs/en/jest-object#jestclearallmocks)

**Review by rflechtner — CHANGES_REQUESTED** [review]
the failing integration tests indicate that this broke the API by requiring that a KILT address be sent to the faucet endpoint. You basically have two options to fix this:
1. translate a public key to an address before running it through the address validator. In this case it would be cool to amend integration tests so that we call the faucet endpoint with both an address and a public key. Just add an extra test for address.
2. change the api. This would of course require changing the integration tests, but most importantly we'd have to change the `pubkey` key in the payload to `address`; imho we can't ask for a public key but secretly require an address, even in a non-public api. This would mean changing the faucet implementation as well.

File: src/contacts/contacts.module.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (LeonFLK) [reply src/contacts/contacts.module.spec.ts:166]: Ohh yes of course, I missed that I had really the one positive in there as well. This should be split!
I would expand the use of an afterEach to call clearAllMocks().

**Comment by LeonFLK** [comment]
Good points, and yes you are requesting with a pubkey in the Integration tests. I'll talk to TImo, what the services should be capable of. But I think the best solution atm for this would be to convert to address before checking validity and eligibility.

**Comment by LeonFLK** [comment]
Wir wollen nur die addresse erwarten und akzeptieren.

**Review by rflechtner — APPROVED** [review]
Nice! There's an error message to fix (see my comment) but then you're good to go!

File: src/faucet/faucet.controller.ts [hunk 4]
```diff
@@ -29,18 +29,17 @@ export class FaucetController {
   ) {}
 
   @Post('drop')
-  public async drop(@Body('pubkey') pubKey: string, @Req() request: Request) {
-    if (!pubKey) {
+  public async drop(@Body('address') address: string, @Req() request: Request) {
+    if (!address) {
       throw new BadRequestException('no public key')
```
[the remark this answers is not in the document]
Comment (LeonFLK) [reply src/faucet/faucet.controller.ts]: target address sounds good, will push then merge

**Pull request merged**